### PR TITLE
Fbnet benchmark

### PR DESCRIPTION
**Pull request by newstzpz**
## FBNet benchmark
* Added some pre-trained models and benchmark results for FBNet.
* Added a timer to benchmark model inference time in addition to total run time.

We provided pre-trained models for selected FBNet models. 
* All the models are trained from scratched with BN using the training schedule specified below. 
* Evaluation is performed on a single NVIDIA V100 GPU with `MODEL.RPN.POST_NMS_TOP_N_TEST` set to `200`. 

The following inference time is reported:
  * inference total batch=8: Total inference time including data loading, model inference and pre/post preprocessing using 8 images per batch.
  * inference model batch=8: Model inference time only and using 8 images per batch.
  * inference model batch=1: Model inference time only and using 1 image per batch.
  * inferenee caffe2 batch=1: Model inference time for the model in Caffe2 format using 1 image per batch. The Caffe2 models fused the BN to Conv and purely run on C++/CUDA by using Caffe2 ops for rpn/detection post processing.

The pre-trained models are available in the link in the model id.

backbone | type | resolution | inference total batch=8 (s/im) | inference model batch=8 (s/im) | inference model batch=1 (s/im) | inference caffe2 batch=1 (s/im) | box AP | mask AP | model id
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
[R-50-C4](configs/e2e_faster_rcnn_R_50_C4_1x.yaml) (reference) | Fast | 800 | 0.0875 | **0.0793** | 0.0831 | **0.0625** | 34.4 | - | f35857197
[fbnet_chamv1a](configs/e2e_faster_rcnn_fbnet_chamv1a_600.yaml) | Fast | 600 | 0.0315 | **0.0260** | 0.0376 | **0.0188** | 33.5 | - | f100940543
[fbnet_default](configs/e2e_faster_rcnn_fbnet_600.yaml) | Fast | 600 | 0.0316 | **0.0250** | 0.0297 | **0.0130** | 28.2 | - | f101086388
[R-50-C4](configs/e2e_mask_rcnn_R_50_C4_1x.yaml) (reference) | Mask | 800 | 0.0918 | **0.0848** | 0.0844 | - | 35.2 | 31.0 | f35858791
[fbnet_xirb16d](configs/e2e_mask_rcnn_fbnet_xirb16d_dsmask_600.yaml) | Mask | 600 | 0.0386 | **0.0319** | 0.0356 | - | 30.7 | 26.9 | f101086394
[fbnet_default](configs/e2e_mask_rcnn_fbnet_600.yaml) | Mask | 600 | 0.0327 | **0.0269** | 0.0385 | - | 29.0 | 26.1 | f101086385

### CODE REVIEW

File: maskrcnn_benchmark/engine/inference.py [hunk 1]
```diff
@@ -11,17 +11,22 @@
 from ..utils.comm import is_main_process
```
Comment (fmassa) [inline maskrcnn_benchmark/engine/inference.py]: I think we can now remove `import datetime` from this file?

File: maskrcnn_benchmark/engine/inference.py [hunk 2]
```diff
@@ -11,17 +11,22 @@
...
-    for i, batch in enumerate(tqdm(data_loader)):
+    for _, batch in enumerate(tqdm(data_loader)):
         images, targets, image_ids = batch
         images = images.to(device)
         with torch.no_grad():
+            if timer:
+                timer.tic()
             output = model(images)
+            if timer:
+                timer.toc()
```
Comment (fmassa) [inline maskrcnn_benchmark/engine/inference.py]: I'm afraid we might need to add a `torch.cuda.synchronize()` here to precisely measure the GPU time, or else this will be inexact because CUDA operations are asynchronous

File: tests/test_detectors.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    "e2e_mask_rcnn_fbnet.yaml",
+
+    # keypoints
+    # TODO: fail to run for random model due to empty head input
+    # "e2e_keypoint_rcnn_R_50_FPN_1x.yaml",
+
+    # gn
+    "gn_baselines/e2e_faster_rcnn_R_50_FPN_1x_gn.yaml",
+    # TODO: fail to run for random model due to empty head input
+    # "gn_baselines/e2e_mask_rcnn_R_50_FPN_Xconv1fc_1x_gn.yaml",
```
Comment (fmassa) [inline tests/test_detectors.py]: Could you expand on why those two cases fail?

File: MODEL_ZOO.md [hunk 4]
```diff
@@ -33,7 +33,28 @@ backbone | type | lr sched | im / gpu | train mem(GB) | train time (s/iter) | to
...
+
+The following inference time is reported:
+  * inference total batch=8: Total inference time including data loading, model inference and pre/post preprocessing using 8 images per batch.
+  * inference model batch=8: Model inference time only and using 8 images per batch.
+  * inference model batch=1: Model inference time only and using 1 image per batch.
+  * inferenee caffe2 batch=1: Model inference time for the model in Caffe2 format using 1 image per batch. The Caffe2 models fused the BN to Conv and purely run on C++/CUDA by using Caffe2 ops for rpn/detection post processing.
+
+The pre-trained models are available in the link in the model id.
+
+backbone | type | resolution | lr sched | im / gpu | train mem(GB) | train time (s/iter) | total train time (hr) | inference total batch=8 (s/im) | inference model batch=8 (s/im) | inference model batch=1 (s/im) | inference caffe2 batch=1 (s/im) | box AP | mask AP | model id
```
Comment (fmassa) [inline MODEL_ZOO.md]: There seems to be a small problem with the table formatting: seeing the diff in rich format doesn't show a table but the raw data

File: maskrcnn_benchmark/engine/inference.py (hunk 2 above)
Comment (newstzpz) [reply maskrcnn_benchmark/engine/inference.py]: I will rerun the benchmark with synchronize. But I guess it may not affect the results as there are post processing steps in python code which runs after the CUDA operations. So I would expect all the computation should have finished at that time?

Comment (fmassa) [reply maskrcnn_benchmark/engine/inference.py]: Maybe. Some post-processing is running on the GPU though, so to be safe it might be better to synchronize anyway.

Comment (newstzpz) [reply maskrcnn_benchmark/engine/inference.py]: I added the synchronize() and reran all the evaluations. It seems the there is not much influence for the bbox models (some are slower, some are faster), but the mask models is in general a bit slower. I have updated the numbers in the document.

File: tests/test_detectors.py (hunk 3 above)
Comment (newstzpz) [reply tests/test_detectors.py]: It seemed to fail when no input to the keypoint head. The error message is follow:
```
  File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/roi_heads.py", line 53, in forward
    x, detections, loss_keypoint = self.keypoint(keypoint_features, detections, targets)
  File "/anaconda3/envs/maskrcnn_benchmark/lib/python3.7/site-packages/torch/nn/modules/module.py", line 491, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/keypoint_head/keypoint_head.py", line 42, in forward
    result = self.post_processor(kp_logits, proposals)
  File "/anaconda3/envs/maskrcnn_benchmark/lib/python3.7/site-packages/torch/nn/modules/module.py", line 491, in __call__
    result = self.forward(*input, **kwargs)
  File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/keypoint_head/inference.py", line 15, in forward
    mask_prob, scores = self.keypointer(x, boxes)
  File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/keypoint_head/inference.py", line 114, in __call__
    assert len(boxes) == 1
AssertionError
```

Comment (fmassa) [reply tests/test_detectors.py]: Ok, thanks!
I'll see if I fix this in the future

**Comment by zimenglan-sysu-512** [comment]
i have a suggestion: make a folder `configs/fbnet` and pull the corresponding yaml files into it.

**Comment by jario-jin** [comment]
I have a question, why does Fbnet not use FPN? Would it be ineffective to detect small objects?

**Comment by fmassa** [comment]
I've updated the README with the pre-trained models for Detection.
If you want to use it, just download it following the link on the MODEL_ZOO, and pass the `MODEL.WEIGHT` to point to the downloaded file.

Thanks a lot @newstzpz !

File: tests/test_detectors.py (hunk 3 above)
Comment (rwwneu) [reply tests/test_detectors.py:31]: > It seemed to fail when no input to the keypoint head. The error message is follow:
> 
> ```
>   File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/roi_heads.py", line 53, in forward
>     x, detections, loss_keypoint = self.keypoint(keypoint_features, detections, targets)
>   File "/anaconda3/envs/maskrcnn_benchmark/lib/python3.7/site-packages/torch/nn/modules/module.py", line 491, in __call__
>     result = self.forward(*input, **kwargs)
>   File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/keypoint_head/keypoint_head.py", line 42, in forward
>     result = self.post_processor(kp_logits, proposals)
>   File "/anaconda3/envs/maskrcnn_benchmark/lib/python3.7/site-packages/torch/nn/modules/module.py", line 491, in __call__
>     result = self.forward(*input, **kwargs)
>   File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/keypoint_head/inference.py", line 15, in forward
>     mask_prob, scores = self.keypointer(x, boxes)
>   File "/Users/stzpz/code/newstzpz/forked/maskrcnn-benchmark/maskrcnn_benchmark/modeling/roi_heads/keypoint_head/inference.py", line 114, in __call__
>     assert len(boxes) == 1
> AssertionError
> ```
Can you tell me how to solve this problem? We look forward to your reply.thanks a lot
